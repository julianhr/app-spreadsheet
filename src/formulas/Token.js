class Token {
  constructor(type, text, whitespace='') {
    this.type = type
    this.text = text
    this.value = null
    this.html = null
    this.whitespace = whitespace
  }

  _repr() {
    return Object
      .entries(this)
      .reduce((prev, curr) => {
        if (curr[1] !== null) {
          prev[curr[0]] = curr[1]
        }

        return prev
      }, {})
  }
}

export default Token
