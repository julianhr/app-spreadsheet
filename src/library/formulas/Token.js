class Token {
  constructor(type, text, whitespace='', value=null) {
    this.type = type
    this.text = text
    this.value = value
    this.whitespace = whitespace
  }

  _repr() {
    const { text, value, type, whitespace } = this
    return { text, value, type, whitespace }
  }
}

export default Token
