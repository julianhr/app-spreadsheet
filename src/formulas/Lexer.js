import Token from './Token'

const TOKENS = {
  // internal
  EOF: 'EOF',
  UNKNOWN: 'UNKNOWN',
  // structure
  EQUALS: 'EQUALS',
  COMMA: 'COMMA',
  COLON: 'COLON',
  LPAREN: 'LPAREN',
  RPAREN: 'RPAREN',
  // operators
  PLUS: 'PLUS',
  MINUS: 'MINUS',
  MULT: 'MULT',
  DIV: 'DIV',
  // entities
  NUMBER: 'NUMBER',
  CELL: 'CELL',
  FUNCTION: 'FUNCTION',
}

const t = TOKENS

class Rule {
  constructor(regex, groupIndex, token) {
    this.regex = regex
    this.groupIndex = groupIndex
    this.token = token
  }

  setIndex(index) {
    this.regex.lastIndex = index
  }

  test(input) {
    return this.regex.exec(input)
  }
}

const GRAMMAR = [
  // internal
  new Rule(/=/g, 0, t.EQUALS),
  new Rule(/,/g, 0, t.COMMA),
  new Rule(/:/g, 0, t.COLON),
  new Rule(/\(/g, 0, t.LPAREN),
  new Rule(/\)/g, 0, t.RPAREN),
  // operators
  new Rule(/\+/g, 0, t.PLUS),
  new Rule(/\-/g, 0, t.MINUS), // eslint-disable-line
  new Rule(/\*/g, 0, t.MULT),
  new Rule(/\//g, 0, t.DIV),
  // multi-character
  new Rule(/[\d\.]+/g, 0, t.NUMBER), // eslint-disable-line
  new Rule(/[a-z]+[\d]+/gi, 0, t.CELL),
  new Rule(/([a-z]+)\(/gi, 1, t.FUNCTION),
]


class Lexer {
  constructor(input, grammar) {
    this.input = input
    this.grammar = grammar
    this.index = 0
    this.char = input[this.index]
    this.markers = []
    this.tokens = []
  }

  getTokens() {
    while (this.char !== t.EOF) {
      const token = this.nextToken()
      this.tokens.push(token)
    }

    return this.tokens
  }

  nextToken() {
    if (this.char === t.EOF) {
      return new Token(t.EOF, t.EOF)
    }

    // console.log('before whitespace', this.char, this.index)
    const whitespace = this.getWhitespace()
    let token

    for (let rule of this.grammar) {
      rule.setIndex(this.index)
      const match = rule.test(this.input)

      if (match && match.index === this.index) {
        const text = match[rule.groupIndex]
        token = new Token(rule.token, text, whitespace)
        this.index += text.length - 1
        this.consume()
        break
      }
    }

    if (!token) {
      token = this.tokenUNKNOWN(whitespace)
    }

    return token
  }

  tokenUNKNOWN(whitespace) {
    const chars = []

    while (!(this.isEOF() || this.isSeparator() || this.isWhitespace())) {
      chars.push(this.char)
      this.consume()
    }
    
    const word = chars.join('')
    return new Token(t.UNKNOWN, word, whitespace)
  }

  consume() {
    this.index++

    if (this.index < this.input.length) {
      this.char = this.input[this.index]
    } else {
      this.char = t.EOF
    }
  }

  isEOF() {
    return this.char === t.EOF
  }

  isWhitespace() {
    return !this.isEOF() && Boolean(this.char.match(/\s/))
  }

  isSeparator() {
    return !this.isEOF() && Boolean(this.char.match(/[()+-/*]{1}/))
  }

  getWhitespace() {
    let count = 0

    while (this.isWhitespace()) {
      count++
      this.consume()
    }

    return Array(count).fill(' ').join('')
  }
}

export default Lexer
export { Rule, GRAMMAR, TOKENS, Lexer }
