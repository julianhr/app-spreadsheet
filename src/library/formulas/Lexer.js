import Token from './Token'
import formulas from './formulas'


const SIMPLE_SYMBOLS = {
  // structure
  '=': 'EQUALS',
  ',': 'COMMA',
  '(': 'LPAREN',
  ')': 'RPAREN',
  ':': 'COLON',
  // operators
  '+': 'PLUS',
  '-': 'MINUS',
  '*': 'MULT',
  '/': 'DIV',
}

class Lexer {
  // TOKENS
  EOF = '<EOF>'
  // complex symbols
  NUMBER = 'NUMBER'
  CELL = 'CELL'
  FUNCTION = 'FUNCTION'

  constructor(input) {
    this.pointer = 0
    this.input = input
    this.char = input[this.pointer]
    this.markers = []
    this.memo = {}
    this.setSimpleTokens()
  }

  setSimpleTokens() {
    for (let token of Object.values(SIMPLE_SYMBOLS)) {
      this[token] = token
    }
  }

  nextToken() {
    let whitespace

    if (this.char === this.EOF) {
      return new Token(this.EOF, this.EOF)
    }

    whitespace = this.getWhitespace()

    // simple symbols
    if (this.isSimpleSymbol()) {
      return this.tokenSimpleSymbol(whitespace)
      // complex symbols
    } else if (this.isNumberToken()) {
      return this.tokenNUMBER(whitespace)
    } else if (this.isCellToken()) {
      return this.tokenCELL(whitespace)
    } else if (this.isFunctionToken()) {
      return this.tokenFUNCTION(whitespace)
    } else {
      throw new Error(`Invalid character: ${this.char}`)
    }
  }

  consume() {
    this.pointer++

    if (this.pointer >= this.input.length) {
      this.char = this.EOF
    } else {
      this.char = this.input[this.pointer]
    }
  }

  match(regexFn) {
    const regex = regexFn()
    regex.lastIndex = this.pointer
    const reMatch = regex.exec(this.input)

    if (reMatch && reMatch.index === this.pointer) {
      this.pointer += reMatch[0].length - 1
      this.consume()
    } else {
      throw new Error(`No match for ${regexFn.name} at position ${this.pointer}`)
    }
  }

  lparen() {
    return new RegExp('\\(', 'g')
  }

  digit() {
    return new RegExp('\\d', 'g')
  }

  letter() {
    return new RegExp('[a-z]', 'gi')
  }

  letters() {
    return new RegExp('[a-z]+', 'ig')
  }

  number() {
    const reFloat = '\\d*\\.\\d+'
    const reInteger = '\\d+\\.?'
    return new RegExp(`${reFloat}|${reInteger}`, 'g') // float or integer
  }

  digits() {
    return new RegExp('\\d+', 'g')
  }

  tokenSimpleSymbol(whitespace) {
    const token = new Token(SIMPLE_SYMBOLS[this.char], this.char, whitespace)
    this.consume()
    return token
  }

  isSimpleSymbol() {
    return this.char in SIMPLE_SYMBOLS
  }

  isNumberToken() {
    let isMatch
    this.markPointer()

    try {
      this.match(this.number)
      if (this.char === '.') { throw new Error('Number has more than one dot') }
      isMatch = true
    } catch (error) {
      this.restorePointer()
      isMatch = false
    }

    return isMatch
  }

  tokenNUMBER(whitespace) {
    const start = this.markers.pop()
    let numStr = this.input.slice(start, this.pointer)
    const num = parseFloat(numStr)

    if (numStr[numStr.length - 1] === '.') {
      numStr = numStr.slice(0, -1)
    }

    return new Token(this.NUMBER, numStr, whitespace, num)
  }

  isCellToken() {
    let isMatch

    this.markPointer()

    try {
      this.match(this.letters)
      this.match(this.digits)
      isMatch = true
    } catch (error) {
      isMatch = false
    }

    this.restorePointer()
    return isMatch
  }

  tokenCELL(whitespace) {
    this.markPointer()
    this.match(this.letters)
    const letters = this.seekSlice(1).toUpperCase()

    this.markPointer()
    this.match(this.digits)
    const digits = this.seekSlice(1)

    const cellName = letters + digits
    const cellValue = `${letters}-${digits}`
    this.clearMarkers()
    return new Token(this.CELL, cellName, whitespace, cellValue)
  }

  isFunctionToken() {
    let isMatch, fnName

    this.markPointer()

    try {
      this.match(this.letters)
      this.markPointer()
      this.match(this.lparen)
      fnName = this.seekSlice(2, 1).toUpperCase()
      isMatch = fnName in formulas
    } catch (error) {
      isMatch = false
    }

    if (isMatch) {
      // subtract LPARENT
      this.memo[this.seekPointer(2)] = this.pointer - 2
    }

    this.restorePointer(2)
    return isMatch
  }

  tokenFUNCTION(whitespace) {
    const endI = this.memo[this.pointer]
    const fnName = this.input.slice(this.pointer, endI + 1).toUpperCase()
    this.pointer += fnName.length - 1
    this.consume()
    return new Token(this.FUNCTION, fnName, whitespace)
  }

  isWhitespace() {
    return Boolean(this.char.match(/\s/))
  }

  getWhitespace() {
    let len = 0

    while (this.isWhitespace(this.char)) {
      len += 1
      this.consume()
    }

    return Array(len)
      .fill(' ')
      .join('')
  }

  markPointer() {
    this.markers.push(this.pointer)
  }

  clearMarkers() {
    this.markers = []
  }

  restorePointer(steps=1) {
    for (let i = 0; i < steps; i++) {
      this.pointer = this.markers.pop()
    }
  }

  seekPointer(steps) {
    if (steps > this.markers.length) {
      throw new Error('step outside of range')
    }

    const index = this.markers.length - steps
    return this.markers[index]
  }

  seekSlice(start, end) {
    let startI, endI

    if (start > this.markers.length || end > this.markers.length) {
      throw new Error('Seek steps are outside range.')
    }

    startI = this.markers[this.markers.length - start]

    if (end === undefined) {
      return this.input.slice(startI, this.pointer)
    } else {
      endI = this.markers[this.markers.length - end]
      return this.input.slice(startI, endI)
    }

  }
}

export default Lexer
export { SIMPLE_SYMBOLS }
