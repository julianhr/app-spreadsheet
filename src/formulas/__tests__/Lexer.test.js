import Token from '../Token'
import { Lexer, Rule, GRAMMAR, TOKENS as t } from '../Lexer'


const toBeWithMsg = {
  toBeWithMsg: (received, expected) => {
    const pass = received.result === expected

    return {
      pass,
      message: () => `context: "${received.context}", expected: ${expected}`
    }
  }
}

describe('Rule', () => {
  it('instantiates correctly', () => {
    const regex = /[a-z]+/gi
    const groupIndex = 0
    const token = 'TEST'
    const rule = new Rule(regex, groupIndex, token)

    expect(rule.regex).toEqual(regex)
    expect(rule.groupIndex).toEqual(groupIndex)
    expect(rule.token).toEqual(token)
  })

  it('sets regex last index', () => {
    const rule = new Rule(/[a-z]+/gi, 0, 'TEST')
    rule.setIndex(10)
    expect(rule.regex.lastIndex).toEqual(10)
  })

  it('executes regex match', () => {
    const input = '1234 test 5678'
    const rule = new Rule(/[a-z]+/gi, 0, 'TEST')
    const match = rule.test(input)
    expect(match.index).toEqual(5)
    expect(match[0]).toEqual('test')
  })
})

describe('Lexer', () => {
  describe('#consume', () => {
    it('increases index if index < input length and sets char', () => {
      const lexer = new Lexer('test')
      lexer.consume()
      expect(lexer.index).toBe(1)
      expect(lexer.char).toBe('e')
    })

    it('sets EOF if index reaches end of input', () => {
      const lexer = new Lexer('hi')
      lexer.consume()
      lexer.consume()
      expect(lexer.index).toBe(2)
      expect(lexer.char).toBe(t.EOF)

      lexer.consume()
      expect(lexer.index).toBe(3)
      expect(lexer.char).toBe(t.EOF)
    })
  })

  test('#isEOF', () => {
    let input, lexer

    input = '=5'
    lexer = new Lexer(input)
    expect(lexer.isEOF()).toBe(false)
    Array(2).fill('').forEach( _ => lexer.consume() ) // consume all tokens
    expect(lexer.isEOF()).toBe(true)
    lexer.nextToken()
    expect(lexer.isEOF()).toBe(true)
    
    input = ''
    lexer = new Lexer(input)
    Array(1).fill('').forEach( _ => lexer.consume() ) // consume all tokens
    expect(lexer.nextToken().type).toBe(t.EOF)
    expect(lexer.isEOF()).toBe(true)
  })

  test('#isWhitespace', () => {
    const input = '=5 + 10 '
    const lexer = new Lexer(input)
    const cases = [
      false, // =
      false, // 5
      true, // ' '
      false, // +
      true, // ' '
      false, // 1
      false, // 0
      true, // ' '
      false, // EOF
    ]

    expect.extend(toBeWithMsg)

    for (let expected of cases) {
      const received = {
        context: lexer.char,
        result: lexer.isWhitespace(),
      }

      expect(received).toBeWithMsg(expected)
      lexer.consume()
    }
  })

  test('#isSeparator', () => {
    const input = '=5+(8)+P4/3-8*7 +suM(4)'
    const lexer = new Lexer(input)
    const cases = [
      false, // =
      false, // 5
      true, // +
      true, // (
      false, // 8
      true, // )
      true, // +
      false, // P
      false, // 4
      true, // /
      false, // 3
      true, // -
      false, // 8
      true, // *
      false, // 7
      false, // ' '
      true, // +
      false, // s
      false, // u
      false, // M
      true, // (
      false, // 4
      true, // )
      false, // EOF
    ]

    expect.extend(toBeWithMsg)

    for (let expected of cases) {
      const received = {
        context: lexer.char,
        result: lexer.isSeparator(),
      }

      expect(received).toBeWithMsg(expected)
      lexer.consume()
    }
  })

  test('#getWhitespace', () => {
    const input = '= 5  +10 '
    const lexer = new Lexer(input)

    lexer.consume()
    expect(lexer.getWhitespace()).toBe(' ')
    lexer.consume()
    expect(lexer.getWhitespace()).toBe('  ')
    lexer.consume()
    expect(lexer.getWhitespace()).toBe('')
  })

  test('token EOF', () => {
    const input = '=5'
    const lexer = new Lexer(input)
    let token, expected

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // consume all tokens
    expected = { type: t.EOF, text: t.EOF, whitespace: '' }
    token = lexer.nextToken()
    expect(token._repr()).toEqual(expected)

    token = lexer.nextToken()
    expect(token._repr()).toEqual(expected)
  })

  test('token UNKNOWN', () => {
    const input = '=ñu /( ?invalid)'
    const lexer = new Lexer(input)
    let token, expected

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // consume all tokens
    expected = { type: t.UNKNOWN, text: 'ñu', whitespace: '' }
    token = lexer.nextToken()
    expect(token._repr()).toEqual(expected)
    
    Array(2).fill('').forEach(_ => lexer.nextToken() ) // consume all tokens
    expected = { type: t.UNKNOWN, text: '?invalid', whitespace: ' ' }
    token = lexer.nextToken()
    expect(token._repr()).toEqual(expected)
  })

  test('token TEXT', () => {
    const cases = [
      '  ',
      '.',
      'hi',
      ' hi  ',
      '+1',
      ' 1',
      ' =2',
      ' =2+3',
    ]

    for (let input of cases) {
      const lexer = new Lexer(input)
      const expected = { type: t.TEXT, text: input, whitespace: '' }
      const tokens = lexer.getTokens()

      expect(tokens.length).toBe(1)
      expect(tokens[0]._repr()).toEqual(expected)
    }
  })

  test('token EQUALS', () => {
    const cases = [
      '=5+8',
      '= 5 +8 ',
    ]

    cases.forEach(input => {
      let lexer = new Lexer(input)
      const expected = { type: t.EQUALS, text: '=', whitespace: '' }
      const token = lexer.nextToken()
      expect(token._repr()).toEqual(expected)
    })
  })

  test('token COMMA', () => {
    let token, expected
    const input = '=SUM(5  ,6 ,7,8)'
    const lexer = new Lexer(input)

    Array(4).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.COMMA, text: ',', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.COMMA, text: ',', whitespace: ' ' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.COMMA, text: ',', whitespace: '' }
    expect(token._repr()).toEqual(expected)
  })

  test('token LPAREN', () => {
    let token, expected
    const input = '=SUM(5) + (6+  (7))'
    const lexer = new Lexer(input)

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.LPAREN, text: '(', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(3).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.LPAREN, text: '(', whitespace: ' ' }
    expect(token._repr()).toEqual(expected)

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.LPAREN, text: '(', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token RPAREN', () => {
    const input = '=(5) + 8+(5  ) +8'
    const lexer = new Lexer(input)
    let token, expected

    Array(3).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.RPAREN, text: ')', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(5).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.RPAREN, text: ')', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token COLON', () => {
    let token, expected
    const input = '=SUM(B2:D4)*sum(BC45  : Z5)'
    let lexer = new Lexer(input)

    Array(4).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.COLON, text: ':', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(6).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.COLON, text: ':', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token PLUS', () => {
    const input = '=5+3  + 3'
    const lexer = new Lexer(input)
    let token, expected

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.PLUS, text: '+', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.PLUS, text: '+', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token MINUS', () => {
    const input = '=5-3  - 7'
    const lexer = new Lexer(input)
    let token, expected

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.MINUS, text: '-', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.MINUS, text: '-', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token DIV', () => {
    const input = '=5/3  / 7'
    const lexer = new Lexer(input)
    let token, expected

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.DIV, text: '/', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.DIV, text: '/', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token MULT', () => {
    const input = '=5*3  * 8'
    const lexer = new Lexer(input)
    let token, expected

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.MULT, text: '*', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.MULT, text: '*', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  describe('token NUMBER', () => {
    it('returns 0 if string is empty', () => {
      const input = ''
      const lexer = new Lexer(input)
      const expected = { type: t.NUMBER, text: '0', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)
    })

    it('returns digit if string is a number', () => {
      const input = '47'
      const lexer = new Lexer(input)
      const expected = { type: t.NUMBER, text: '47', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)
    })

    it('parses integers', () => {
      const input = '=5 + 58 /(  23)'
      const lexer = new Lexer(input)
      let expected

      Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '5', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '58', whitespace: ' ' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '23', whitespace: '  ' }
      expect(lexer.nextToken()._repr()).toEqual(expected)
    })

    it('parses floats', () => {
      let expected
      const input = '=5. + 58.3 /(  .23)'
      const lexer = new Lexer(input)
      Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '5.', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '58.3', whitespace: ' ' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '.23', whitespace: '  ' }
      expect(lexer.nextToken()._repr()).toEqual(expected)
    })

    it('includes malformed float numbers', () => {
      let expected
      const input = '=0.13.2.6+.9.+..5'
      const lexer = new Lexer(input)

      Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
      expected = { type: t.NUMBER, text: '0.13.2.6', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      expected = { type: t.PLUS, text: '+', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      expected = { type: t.NUMBER, text: '.9.', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      expected = { type: t.PLUS, text: '+', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)

      expected = { type: t.NUMBER, text: '..5', whitespace: '' }
      expect(lexer.nextToken()._repr()).toEqual(expected)
    })
  })

  test('token CELL', () => {
    const input = '=5+B2 +  c3/(P3* cA25)'
    const lexer = new Lexer(input)
    let expected, token

    Array(3).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.CELL, text: 'B2', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.CELL, text: 'c3', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)

    Array(2).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.CELL, text: 'P3', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.CELL, text: 'cA25', whitespace: ' ' }
    expect(token._repr()).toEqual(expected)
  })

  test('token FUNCTION', () => {
    const input = '=5+SUM( sUm( 5)/  sum(5,8,9))'
    const lexer = new Lexer(input)
    let expected, token

    Array(3).fill('').forEach(_ => lexer.nextToken()) // skip to token
    token = lexer.nextToken()
    expected = { type: t.FUNCTION, text: 'SUM', whitespace: '' }
    expect(token._repr()).toEqual(expected)

    Array(1).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.FUNCTION, text: 'sUm', whitespace: ' ' }
    expect(token._repr()).toEqual(expected)

    Array(4).fill('').forEach(_ => lexer.nextToken() ) // skip to token
    token = lexer.nextToken()
    expected = { type: t.FUNCTION, text: 'sum', whitespace: '  ' }
    expect(token._repr()).toEqual(expected)
  })

  describe('#getTokens', () => {
    test('', () => {
      const input = ''
      const lexer = new Lexer(input)
      const tokens = lexer.getTokens()
      const token = new Token(t.NUMBER, '0', '')

      expect(tokens.length).toBe(1)
      expect(tokens[0]._repr()).toEqual(token._repr())
    })

    test('=2', () => {
      const input = '=2'
      const lexer = new Lexer(input)
      const tokens = lexer.getTokens()
      const expected = [
        new Token(t.EQUALS, '=', ''),
        new Token(t.NUMBER, '2', ''),
      ]

      expected.forEach((token, i) => {
        expect(tokens[i]._repr()).toEqual(token._repr())
      })
    })

    test('=2 *  (4+5 )', () => {
      const input = '=2 *  (4+5 ) '
      const lexer = new Lexer(input)
      const tokens = lexer.getTokens()
      const expected = [
        new Token(t.EQUALS, '=', ''),
        new Token(t.NUMBER, '2', ''),
        new Token(t.MULT, '*', ' '),
        new Token(t.LPAREN, '(', '  '),
        new Token(t.NUMBER, '4', ''),
        new Token(t.PLUS, '+', ''),
        new Token(t.NUMBER, '5', ''),
        new Token(t.RPAREN, ')', ' '),
      ]

      expected.forEach((token, i) => {
        expect(token._repr()).toEqual(tokens[i]._repr())
      })
    })

    test('=5+ 2.5/(9.25.)+sum( .84.5, 4)', () => {
      const input = '=5+ 2.5/(9.25.)+sum( .84.5, 4)'
      const lexer = new Lexer(input)
      const expected = [
        new Token(t.EQUALS, '=', ''),
        new Token(t.NUMBER, '5', ''),
        new Token(t.PLUS, '+', ''),
        new Token(t.NUMBER, '2.5', ' '),
        new Token(t.DIV, '/', ''),
        new Token(t.LPAREN, '(', ''),
        new Token(t.NUMBER, '9.25.', ''),
        new Token(t.RPAREN, ')', ''),
        new Token(t.PLUS, '+', ''),
        new Token(t.FUNCTION, 'sum', ''),
        new Token(t.LPAREN, '(', ''),
        new Token(t.NUMBER, '.84.5', ' '),
        new Token(t.COMMA, ',', ''),
        new Token(t.NUMBER, '4', ' '),
        new Token(t.RPAREN, ')', ''),
      ]
      const tokens = lexer.getTokens()

      expect(tokens.length).toEqual(expected.length)

      expected.forEach((token, i) => {
        expect(tokens[i]._repr()).toEqual(token._repr())
      })
    })
  })
})
