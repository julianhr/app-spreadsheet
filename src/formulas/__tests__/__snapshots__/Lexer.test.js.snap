// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer #getTokens  1`] = `
Token {
  "category": "entity",
  "html": null,
  "index": -1,
  "text": "0",
  "type": "NUMBER",
  "value": 0,
  "whitespace": 0,
}
`;

exports[`Lexer token NUMBER includes malformed float numbers 1`] = `
Token {
  "category": "entity",
  "html": null,
  "index": 14,
  "text": "..5",
  "type": "TEXT",
  "value": null,
  "whitespace": 0,
}
`;

exports[`Lexer token TEXT matches snapshot 1`] = `
Token {
  "category": "entity",
  "html": null,
  "index": 0,
  "text": "test input",
  "type": "TEXT",
  "value": null,
  "whitespace": 0,
}
`;
